---
title: "Gather min P values per position from all combined data"
author: "Lucas Moitinho-Silva"
date: "`r Sys.Date()`"
output:
   BiocStyle::html_document:
      toc: true
      df_print: paged
      self_contained: true
      code_download: true
      highlight: tango
#bibliography: mylib.bib
editor_options:
  chunk_output_type: console
params:
  FIGPATH: "figures/"
  d.out: "./"
---


# Introduction


# Preparations

## Set global options

```{r style, echo = T, results="asis", cache=FALSE, message = F}
#v.01
# Set knit global options
library("knitr")
options(digits = 2, width = 80)
golden_ratio <- (1 + sqrt(5)) / 2
opts_chunk$set(echo = TRUE,
               tidy = FALSE,
               include = TRUE,
               fig.path = params$FIGPATH,
               dev=c("png",'pdf'),
               fig.height = 6,
               fig.width = 5 * golden_ratio,
               comment = '  ',
               dpi = 300,
               cache = FALSE)
library("rmarkdown")

# Pretty outputs
library("BiocStyle")
library("ggthemes") # add ggplot2 themes
library("ggplot2")
theme_set(theme_few(base_size = 14))

# Set seed for reproducibility
set.seed(13)
# Color blind friendly pallet
cbp1 <- c("#E69F00", "#56B4E9", "#009E73",
          "#F0E442", "#0072B2", "#D55E00", "#CC79A7", "#999999")

scale_fill_discrete <- function(...) {
  scale_fill_manual(..., values = cbp1)
}
scale_colour_discrete <- function(...) {
  scale_colour_manual(..., values = cbp1)
}

# Set output directory
d.out <- params$d.out
rm(params)
```


## Load libraries for the session

```{r}
library(phyloseq)
library(tidyverse)
library(data.table)
```


## Set regions
```{r}
beta.dir <- "/work_ifs/sukmb447/projects/skin.mgwas/results/3.meta.analysis/1.microenv/1.beta.diversity/2.gather.and.filter/"
feature.dir <-  "/work_ifs/sukmb447/projects/skin.mgwas/results/3.meta.analysis/1.microenv/2.taxonomic.features/2.gather.and.filter/"
```

# Add tests done and other info

I basically have two types of metanalysis, from beta-diversity and taxonomic features.

```{r}
# Feature indices
feature.indices <-"/work_ifs/sukmb447/projects/skin.mgwas/results/3.meta.analysis/1.microenv/2.taxonomic.features/metasoft/1.metaanalysis/prepare/meta.indices.rds" %>% 
  readRDS()

#  Beta indicides
micro.indices.in <- "/work_ifs/sukmb447/projects/skin.mgwas/results/1.data.harmonization/3.microbiome.harmonized/phyloseq.rds" %>% 
  readRDS() %>% 
  sample_data() %>%
  data.frame() %>% 
  select(microenv, index) %>% 
  unique()
beta.indices <- NULL
for (i in micro.indices.in %>% 
     pull(microenv) %>% 
     unique()){
  df <- micro.indices.in %>% filter(microenv == i)
  beta.indices  <- bind_rows(beta.indices, data.frame(microenv = i,
                                                      indices = df %>% pull(index) %>% paste(collapse = "")))
}
rm(micro.indices.in, df)

# SNPs
snps <- "/work_ifs/sukmb447/projects/skin.mgwas/kora.popgen.pca/PopGen_and_KORA.harmonized.all.bim" %>% 
  fread(header = F) %>% 
  .[,2]
colnames(snps) <- "snp.name"
snps$P <- 1
snps$metaid.type <- ""

```

# Function to read files
```{r}
#type <- "feature"
#i <- 178 # feature Sebaceous

read.files.get.p <- function(i, type){
  # Get microenvironment data
  if(type == "beta"){microenv <- beta.indices[i,] %>% pull(microenv)}
  if(type == "feature"){microenv <- feature.indices[i,] %>% pull(microenv)}
  
  # Get metaanalysis
  if(microenv == "Dry"){metanalysis = "comb"}
  if(microenv %in% c("Moist", "Sebaceous")){metanalysis <- "meta"}
  
  # Core files
  if(type == "beta"){
    meta.id <- beta.indices[i,] %>% pull(indices)
    file <- paste0(beta.dir, "gathered.meta", meta.id,".tsv.gz")
  }
  if(type == "feature"){
    meta.id <- feature.indices[i,] %>% pull(meta.id)
    test.1 <- feature.indices[i,] %>% pull(test.1)
    test.2 <- feature.indices[i,] %>% pull(test.2)
    index.1 <- feature.indices[i,] %>% pull(index.1)
    index.2 <- feature.indices[i,] %>% pull(index.2)
    file <- paste0(feature.dir, "parsedmeta", meta.id,".test", test.1, ".test",test.2, ".index", index.1, ".index", index.2, ".txt.gz")
  }
  
  # Format file
  df <- fread(file) %>% 
    select(snp.name, P.meta)
  colnames(df) <- c("snp.name", "P")
  
  # Add meta.id and type
  df$meta.id <- meta.id
  df$type <- type
  
  # Output
  
  return(df)
}
```


# Read all files

```{r}

for (i in 1:nrow(beta.indices)){
  tryCatch({
    
    df <- read.files.get.p(i, "beta")
    snps <- left_join(snps, df, by = "snp.name")
    # select minimum P
    snps$P <- with(snps, pmin(P.x, P.y, na.rm = T))
    # annotate where the lowest p value comes from
    snps <- snps %>% 
      mutate(metaid.type = case_when(P == P.x & P != P.y ~ metaid.type,
                                     P == P.x & P == P.y ~ paste0(metaid.type,
                                                                 ",", 
                                                                 paste0(meta.id, type)),
                                     P != P.x & P == P.y ~ paste0(meta.id, type),
                                     TRUE ~ metaid.type))
    #Select final xolumna
    snps <- select(snps, snp.name, P, metaid.type)

    },
    error = function(x){
      print(i)})
}


for (i in 1:nrow(feature.indices)){
  tryCatch({
    
    df <- read.files.get.p(i, "feature")
    snps <- left_join(snps, df, by = "snp.name")
    # select minimum P
    snps$P <- with(snps, pmin(P.x, P.y, na.rm = T))
    # annotate where the lowest p value comes from
    snps <- snps %>% 
      mutate(metaid.type = case_when(P == P.x & P != P.y ~ metaid.type,
                                     P == P.x & P == P.y ~ paste0(metaid.type,
                                                                 ",", 
                                                                 paste0(meta.id, type)),
                                     P != P.x & P == P.y ~ paste0(meta.id, type),
                                     TRUE ~ metaid.type))
    #Select final xolumna
    snps <- select(snps, snp.name, P, metaid.type)
    
    },
    error = function(x){
      print(i)})
}

```

# Write output

```{r}
paste0(d.out, "/", "snp.lowestp.rds") %>% 
  saveRDS(snps, .)
```

# Conclusion

# Session information

```{r}
sessionInfo()
```
