---
title: "Make plots"
author: "Lucas Moitinho-Silva"
date: "`r Sys.Date()`"
output:
   BiocStyle::html_document:
      toc: true
      df_print: paged
      self_contained: true
      code_download: true
      highlight: tango
#bibliography: mylib.bib
editor_options:
  chunk_output_type: console
params:
  FIGPATH: "figures/"
  d.out: "./"
---


# Introduction


# Preparations

## Set global options

```{r style, echo = T, results="asis", cache=FALSE, message = F}
#v.01
# Set knit global options
library("knitr")
options(digits = 2, width = 80)
golden_ratio <- (1 + sqrt(5)) / 2
opts_chunk$set(echo = TRUE,
               tidy = FALSE,
               include = TRUE,
               fig.path = params$FIGPATH,
               dev=c("png",'pdf'),
               fig.height = 6,
               fig.width = 5 * golden_ratio,
               comment = '  ',
               dpi = 300,
               cache = FALSE)
library("rmarkdown")

# Pretty outputs
library("BiocStyle")
library("ggthemes") # add ggplot2 themes
library("ggplot2")
theme_set(theme_few())

# Set seed for reproducibility
set.seed(13)
# Color blind friendly pallet
cbp1 <- c("#E69F00", "#56B4E9", "#009E73",
          "#F0E442", "#0072B2", "#D55E00", "#CC79A7", "#999999")

scale_fill_discrete <- function(...) {
  scale_fill_manual(..., values = cbp1)
}
scale_colour_discrete <- function(...) {
  scale_colour_manual(..., values = cbp1)
}

# Set output directory
d.out <- params$d.out
rm(params)
```


## Load libraries for the session

```{r}
library(gdata)
library(grid)
library(gridExtra)
library(gtable)
library(ggplot2)
library(tidyverse)
library(phyloseq)
library(reshape2)
library(ggrepel)
library(data.table)
```

## Import data
```{r}
results <- "~/IKMB/projects/skin.mgwas/results/4.fine.mapping/1.meta/1.microenv/2.annotate.data/combined/lead.snp.bacteria.all.tsv" %>%
  read_tsv() %>% 
  mutate(rsID = if_else(is.na(rsID), 
                        snp.name,
                        rsID))
snps <- "~/IKMB/projects/skin.mgwas/results/5.plot.all.ps/snp.lowestp.rds" %>%
  readRDS()
```

# Get results annotation

```{r}
anno <- results %>% select(microenv, snp.name, rsID, P.meta) %>% unique()

identical(nrow(anno),unique(anno$snp.name) %>% length()) #check for consistency

# tax
tax <- NULL
for (i in unique(anno$snp.name)){
  t <- results %>%
    filter(snp.name == i) %>%
    select(snp.name, name)%>%
    unique() %>%
    pull(name) %>%
    paste(collapse = ";")
  tax <- bind_rows(tax, data.frame(snp.name = i,
                                   tax = t))
}

# genes

gene <- results %>% 
  select(snp.name, genes) %>%
  rename("gene" = "genes") %>% 
  mutate(gene = if_else(is.na(gene),
                        "",
                        gene)) %>% 
  unique()

# Remove Features that do not have a gene name

for (i in 1:nrow(gene)){
 gene[i, "gene"] <-
   gene[i, "gene"] %>% 
    str_split(",") %>% 
    .[[1]] %>% 
    .[!str_detect(., "^ENSG00")] %>%
    .[!str_detect(., "RNA5SP123")] %>%
    paste(collapse = ",")
}

# add genes and tax to anno
anno <- left_join(anno, tax, by = "snp.name") %>%
  left_join(gene, by = "snp.name") %>%
  rename("P" = "P.meta") 
rm(tax, gene)
```

# Add annotation to the snps
```{r}
snps.no.min <- anti_join(snps, anno, by = c("snp.name")) # Do not merge using P because it may create doubles in case the same snp has a smaller p value from another meta-analysis.
snps.anno <- bind_rows(snps.no.min, anno)
rm(snps.no.min)
```

# Format snps

```{r}
snps.anno <- snps.anno %>%
  separate(snp.name, c("chr", "position"), sep = ":", remove = F)
snps.anno <- snps.anno %>%
  mutate(chr = as.numeric(chr),
         position = as.numeric(position)) %>%
  arrange(chr, position)
```

# Prepare data to plot

```{r}
to.plot <- snps.anno %>% 
  
  # Compute chromosome size
  group_by(chr) %>% 
  summarise(chr_len=max(position)) %>% 
  
  # Calculate cumulative position of each chromosome
  mutate(tot=cumsum(chr_len)-chr_len) %>%
  select(-chr_len) %>%
  
  # Add this info to the initial dataset
  left_join(snps.anno, ., by=c("chr"="chr")) %>%
  
  # Add a cumulative position of each snp.name
  arrange(chr, position) %>%
  mutate( positioncum=position+tot) %>%  # add color 
  mutate(col = if_else((chr %% 2) == 0,
                       "lightgrey", 
                       "darkgrey"))
```

# Prepare labels to plot

Labels will be plotted as a table 

## Add labels to to.plot dataset
```{r}
to.plot <- to.plot %>% 
  mutate(label.key = if_else(is.na(tax),
                             NA_character_,
                             paste(rsID, gene, tax, sep = " : ")),
         label = factor(label.key, levels = unique(label.key)) %>% as.numeric)
```
## Prepare labels data.frame

```{r}
label <- to.plot %>% filter(!is.na(tax)) %>% 
  select(chr,label, label.key) %>%
  mutate(label = paste(label, label.key, sep = " - ")) %>% 
  select(chr, label)

# Legens will be presented in three columns
nchunks <- (nrow(label)/3) %>%  round()
label <- label %>% 
  pull(label) %>% 
  split(., ceiling(seq_along(.)/nchunks))

# Get table with the labels
label.df = cbindX(data.frame(l1 = label[[1]]),
                  data.frame(l2 = label[[2]]),
                  data.frame(l3 = label[[3]]))
label.df[is.na(label.df)] <- ""

## Tweak the aesthetics of the table in the plot
tt <- ttheme_default(base_size = 8, 
                     core=list(fg_params=list(hjust=0, x=0), # adjust to left
                               bg_params = list(
                                 fill = "grey90", # background color
                                 col="grey90"))) # line color
# Make table grob object
gtable <-tableGrob(label.df, rows=NULL, cols = NULL, theme = tt)
# Specify heights and withs of the table
gtable$widths <- unit(c(.36,.29,.32), "npc")
gtable$heights <- unit(rep(0.075, nrow(gtable)), "npc")

```

# Plot figure 
```{r, fig.width= 10, fig.height= 6}
# Prepare X axis
axisdf <- to.plot %>% group_by(chr) %>% summarize(center=( max(positioncum) + min(positioncum) ) / 2 )

p <- ggplot(to.plot, aes(x=positioncum, y=-log10(P), label = label.key)) +
  # Add all points that are not annotated
  geom_point(data = to.plot %>% filter(is.na(label)),
             color= to.plot %>% filter(is.na(label)) %>% .[, col],
             alpha=0.8,
             size=1) +
  # Add annotated points
  geom_point(data = to.plot %>% filter(!is.na(label)), 
             aes(color = microenv),
             alpha=0.8,
             size=2) +
  # custom X axis
  scale_x_continuous(label = axisdf$chr, breaks = axisdf$center) +
  # custom y axis
  ylim(1.8, 12) +
  # Custom the theme:
  theme_bw() +
  # Add label keys
  geom_label_repel(data = to.plot %>% filter(!is.na(label)),
                   aes(label = label),
                   size = 2.2,
                   point.padding=unit(1,'lines'),
                   angle = 90, # vertical names
                   vjust = 0.5,# align
                   direction = "x", # allow to move only at the x axis
                   max.iter = 1e4,
                   force_pull = 0,
                   #Add some line aesthetics
                   segment.curvature = -1e-20, # add a slight curve
                   segment.color = "grey",
                   segment.linetype = 4,
                   min.segment.length = 0, #always draw line
                   # Limit area of labels
                   xlim = c(-Inf, Inf),
                   ylim = c(9.4, Inf)) +
  # Mod the theme
  theme(legend.position = "bottom", 
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        plot.background = element_rect(colour = "darkgrey")) +
  # add cut-offs
  geom_hline(yintercept=-log10(5e-8), linetype="dashed", 
             color = "firebrick", size=0.7) +
  geom_hline(yintercept=-log10(1e-5), linetype="dashed", 
             color = "dodgerblue", size=0.7) +
  # add labels
  labs(x = "Chromossome", y = bquote(-log[10](P[Meta]))) +
  # modify guide names
  guides(color=guide_legend(title="Skin\nmicroenvironment")) +
  # add labels legend
  annotation_custom(gtable,
                    xmin = -Inf,
                    xmax = Inf,
                    ymin = 9.8,
                    ymax = 12.5)
p
```

# Save plot
```{r}
saveRDS(to.plot, paste0(d.out, "/", "plot.df.RDS"))
saveRDS(p, paste0(d.out, "/", "plot.data.RDS"))
```


# Session information

```{r}
sessionInfo()
```
