---
title: "Plot expressions"
author: "Lucas Moitinho-Silva"
date: "`r Sys.Date()`"
output:
   BiocStyle::html_document:
      toc: true
      toc_float: true
      df_print: paged
      self_contained: true
      code_download: true
      highlight: tango
editor_options:
  chunk_output_type: console
params:
  FIGPATH: "figures/"
  d.out: "./"
---
# Preparations

## Set global options

```{r style, echo = T, results="asis", cache=FALSE, message = F}
#v.02
# Set knit global options
library("knitr")
options(digits = 2, width = 80)
golden_ratio <- (1 + sqrt(5)) / 2
opts_chunk$set(echo = TRUE,
               tidy = FALSE,
               include = TRUE,
               fig.path = params$FIGPATH,
               dev=c("png", 'pdf'),
               fig.height = 6,
               fig.width = 5 * golden_ratio,
               comment = '  ',
               dpi = 300,
               cache = FALSE)

# Set ggplot2 them as black and with
library("ggplot2")
library(ggthemes)
theme_set(theme_few(base_size = 16))
# Set seed for reproducibility
set.seed(13)

# # Color blind friendly pallet
cbp1 <- c("#E69F00", "#56B4E9", "#009E73",
          "#F0E442", "#0072B2", "#D55E00", "#CC79A7", "#999999")
# 
# scale_fill_discrete <- function(...) {
#    scale_fill_manual(..., values = cbp1)
# }
# scale_colour_discrete <- function(...) {
#   scale_colour_manual(..., values = cbp1)
# }

# Set output directory
d.out <- params$d.out
rm(params)
```


## Load libraries for the session

```{r}
library(tidyverse)
library(ggplotify)
library(pheatmap)
library(viridis) # color
library(ggrepel)
library(patchwork)
```

## Import data
```{r}
total_expression <- read_tsv("/home/lsilva/IKMB/projects/skin.mgwas/results/4.fine.mapping/1.meta/1.microenv/2.annotate.data/combined/gene.annotation.tsv")
sc_cluster <- readRDS("/home/lsilva/IKMB/projects/skin.mgwas/results/7.sc.transcription/gene.clusters.annotation.rds")
gene_exprs <- "/home/lsilva/IKMB/projects/skin.mgwas/results/7.sc.transcription/gene_exprs.to.plot.rds" %>% 
  readRDS()
de_results <- readRDS("/home/lsilva/IKMB/projects/skin.mgwas/results/10.2d.rnaseq/3.overview/DE.results.rds")
counts <- readRDS("/home/lsilva/IKMB/projects/skin.mgwas/results/10.2d.rnaseq/3.overview/count.data.rds")
pca <- readRDS("/home/lsilva/IKMB/projects/skin.mgwas/results/10.2d.rnaseq/4.PCA/pca.plot.rds")

go <- readRDS("/home/lsilva/IKMB/projects/skin.mgwas/results/10.2d.rnaseq/5.Pathway.enrichment/plot.rds")
genes <- read_tsv("/home/lsilva/IKMB/projects/skin.mgwas/results/4.fine.mapping/1.meta/1.microenv/2.annotate.data/combined/gene.annotation.tsv")
```

# Filter to include only genes detected
```{r}
gene_exprs <- gene_exprs %>% filter(gene_name %in% genes$gene.name)
```


# Plots

## Single-cell clusters

## Total tissue

```{r}
p_total <- total_expression %>% 
  filter(!is.na(NX)) %>% 
  #top_n(8) %>% 
  ggplot(aes(x = fct_reorder(gene.name, desc(NX)), y = NX)) +
  geom_col() +
  ggpubr::rotate_x_text() +
  labs(x = NULL, y = "Normalised\nexpression\nin skin") +
  theme(axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank())
```

## Single-cell averages
```{r}
# Set parameters

gene_levels <- total_expression %>% 
  filter(!is.na(NX)) %>% 
  mutate(levels = fct_reorder(gene.name, desc(NX)) %>% 
           levels()) %>% 
  pull(levels) %>% 
  unique()

# Add cluster info
gene_exprs <- sc_cluster %>% 
  mutate(color = "grey") %>% 
  mutate(broadcell = annotation,
         gene_name = gene) %>% 
  select(gene_name, broadcell, color) %>% 
  distinct() %>% 
  right_join(gene_exprs, by = c("gene_name", "broadcell")) %>% 
  mutate(gene_name = factor(gene_name, levels = gene_levels))

p_exprs <- gene_exprs %>% 
  ggplot(aes(gene_name, broadcell)) +
  geom_tile(fill = gene_exprs$color) + 
  geom_point(aes(fill = avg_log_exprs, size = percent_cells), shape=21, color="black") +
  scale_fill_distiller(
    palette = 'Reds',
    direction = 1,
    name = 'Normalised\nexpression',
    guide = guide_colorbar(frame.colour = "black", ticks.colour = "black")
  ) +
  labs(y = 'Cell type', color = 'Expression', x = 'Gene', size = "Cells (%)") +
  ggpubr::rotate_x_text(45)
```

# Make top plot

```{r, fig.width = 13, fig.height = 7}
p_top <- (p_total / p_exprs) +
  plot_layout(heights = c(1.2, 3.8))
p_top
```

# Make 2D invitro transcriptomics graphs


## PCA

```{r}
pca <- pca +
  guides(col = guide_legend(ncol=1),
         shape = guide_legend(ncol=1)) +
  theme_few(base_size = 16) +
  scale_color_manual(values = cbp1[5:6])
```

# Go terms

```{r}
go <- go +
  theme(legend.position = "right") +
  labs(fill = "Ratio") +
  theme_few(base_size = 16)
  
```


## Gwas genes
```{r}
gwas_genes <- de_results %>% 
  filter(GWAS == "yes") %>%
  ggplot(aes(fct_reorder(Gene_name, log2FoldChange), log2FoldChange, label = Gene_name)) + 
  geom_hline(yintercept = c(-1, 0, 1), linetype = "dashed", color = "grey") +
  geom_errorbar(aes(ymin=log2FoldChange-lfcSE, ymax=log2FoldChange+lfcSE), 
                width=.2,
                position=position_dodge(.9)) +
  geom_point(aes(size = baseMean)) +
  scale_color_manual(values=c("#999999", "#E69F00")) +
  labs(size = "Mean of\nnormalized\ncounts",
       y = "Effect size",
       x = "Gene") +
  coord_flip()
gwas_genes
```


# Make bottom plot

```{r}

bottom_top <- ggpubr::ggarrange(pca, gwas_genes, ncol = 2, widths = c(3, 2), labels = c("a", "c"))

bottom <- ggpubr::ggarrange(bottom_top, go, nrow = 2, labels = c("", "b"))
bottom
```

# Make final plots

```{r, fig.width = 14, fig.height = 6}
p_top
```

```{r, fig.width = 12, fig.height = 12}
bottom
```




# Session information

```{r}
sessionInfo()
```
