#@ batuque: Local laptop

#Cp link to the directory at isilon


#rsync -avh ~/IKMB/projects/skin.mgwas/results/1.data.harmonization/1.microbiome/participant.list sukmb447@medcluster.medfdm.uni-kiel.de:/work_ifs/sukmb447/projects/skin.mgwas/

#@ medcluster

#Load modules
module load miniconda2
#load environment with plink
source activate /work_ifs/sukmb447/apps/conda.envs/gnx # PLINK v1.90b6.12 64-bit (28 Oct 2019)          www.cog-genomics.org/plink/1.9/
#load plink2
module load plink2/2.0-alpha-avx2-20200217


# Set paths
WD="/work_ifs/sukmb447/projects/skin.mgwas"
PREPROCESS=${WD}'/preprocess'
KORA=${PREPROCESS}'/kora'
POPGEN=${PREPROCESS}'/popgen'
KORADATA="/work_ifs/sukmb447/projects/koraff4/genotype/KORA_data_temp"
POPGENDATA="/work_ifs/sukmb447/projects/Popgen/gsa.processing/imputation.pruned"
FOLDERLIST="/work_ifs/sukmb447/projects/skin.mgwas/participant.list"
HARMONIZED=${WD}'/kora.popgen'
PCA=${WD}'/kora.popgen.pca'
MAF=${WD}'/kora.popgen.maf'


#################
###KORA##########
#################

# Preprocess kora
mkdir -p $KORA
cd $KORA

#1. Convert kora to plink. KORA files are impute2, with several chunks for files per chromossome. So, first I need to convert each file idependently to plink. Afterwards, I merge them to have one file per chromosome. Then, I filter for maf, samples and extract imputations that are valid.

#1.1 Convert to plink

## Set variables

## Format samples file. Need to be done because sample file has to have a header.

{ echo 'ID_1'; echo '0'; cat /work_ifs/sukmb447/projects/koraff4/genotype/samples.list/KORAS4F4_N3788_list_of_individuals.txt; } > sample.sample.col1
{ echo 'ID_2'; echo '0'; cat /work_ifs/sukmb447/projects/koraff4/genotype/samples.list/KORAS4F4_N3788_list_of_individuals.txt; } > sample.sample.col2

for i in {1..3788}
do
 echo '0' >> sample.sample.col3a
done

{ echo 'missing'; echo '0'; cat sample.sample.col3a; } > sample.sample.col3
 

paste --delimiters=' ' sample.sample.col1 sample.sample.col2 sample.sample.col3 > sample.sample

rm sample.sample.col*

## I need to extract the information of the chr number and the name (chromossome number + chunk) from the files, e.g. /work_ifs/sukmb447/projects/koraff4/genotype/KORA_data_temp/KORAS4F4_N3788_chr11_4_notmonomorph.impute2.gz . That's why I do some unix jiujitsu. before running the command

for i in ${KORADATA}/*gz
do
 [ -f "$i" ] || continue   # needed to expand the list above using file extension and glob. Continue will exit without error if string returns empy
 echo $i >> log1.convert.to.plink.txt
 # get chromossome and chunk numbers
 z=$i
 z=${z#${KORADATA}/KORAS4F4_N3788_chr}
 z=${z%_notmonomorph.impute2.gz} # code
 echo $z >> log1.convert.to.plink.txt
 
 #get chunk number and edit
 c=${z#[0-9]*_}
 c=$(printf %03d $c) # add zeros, because order of files matter while merging
 echo $c >> log1.convert.to.plink.txt #
 # get chromossome number
 n=${z%_[0-9]*} #chromosome number
 echo $n >> log1.convert.to.plink.txt

 plink --gen $i --sample sample.sample --oxford-single-chr $n --out chr${n}_${c} | tee >> log1.convert.to.plink.txt

unset z
unset n
unset c
done

echo "complete" >> log1.convert.to.plink.txt

## Check for duplicates in variants

cat chr*.bim >> all.bim
cut all.bim -f2 -d " " |sort| uniq -d >> reference.dups1 # something like that
rm all.bim

# No dups, all good

for i in `seq 1 22`;
 do

ls -A1 chr${i}_* | cut -f1 -d "." | sort| uniq >> list.txt # Get sorted name of files to list. because order of files matter while merging

plink --merge-list list.txt --out chr${i}.merged | tee >> log2.merge.chunks.txt # merge them
tar -czvf chr${i}.chunks.tar.gz chr${i}_* list.txt # Store chunks in a single file
rm chr${i}_* list.txt # clean up
 done
 
echo "complete" >> log2.merge.chunks.txt

#1.3 Filter samples & MAF 0.05


paste ${FOLDERLIST}/koraff4.txt ${FOLDERLIST}/koraff4.txt > list.participant # cp list of participants


for i in `seq 1 22`;
 do
plink --bfile chr${i}.merged --keep list.participant --make-bed --out chr${i}.merged.sampfilter | tee >> log3.filter.samples.txt # filter samples
tar --remove-files -czvf chr${i}.merged.tar.gz chr${i}.merged.bed chr${i}.merged.bim chr${i}.merged.fam chr${i}.merged.log # store previous as compressed files
plink --bfile chr${i}.merged.sampfilter --make-bed --out chr${i}.merged.sampfilter.maf005 --maf 0.05 | tee >> log4.filter.maf.txt
tar --remove-files -czvf chr${i}.merged.sampfilter.tar.gz chr${i}.merged.sampfilter.bed chr${i}.merged.sampfilter.bim chr${i}.merged.sampfilter.fam chr${i}.merged.sampfilter.log chr${i}.merged.sampfilter.nosex # store previous as compressed files
 done

 
echo "complete" >> log3.filter.samples.txt
echo "complete" >> log4.filter.maf.txt


###### NEED TO CHECK DOUBLE NAMES IN KORA BEFORE merging 

cat chr*.bim >> all.bim
cut all.bim -f2 -d " " |sort| uniq -d >> reference.dups2 # something like that
rm all.bim


#################
###POPGEN##########
#################

mkdir -p $POPGEN
cd $POPGEN


#1.1 Convert to plink

for i in `seq 1 22` # chromossome x is left behind because there is no availability in KORA dataset
 do
plink --vcf ${POPGENDATA}/chr${i}.imputed.poly.vcf.gz --double-id --out chr${i} | tee >> log1.convert.to.plink.txt
 done
 
echo "complete" >> log1.convert.to.plink.txt


#1.3 Filter samples & MAF 0.05

paste ${FOLDERLIST}/popgen.txt ${FOLDERLIST}/popgen.txt > list.participant # cp list of participants

for i in `seq 1 22`;
 do
plink --bfile chr${i} --keep list.participant --make-bed --out chr${i}.sampfilter | tee >> log3.filter.samples.txt # filter samples
tar --remove-files -czvf chr${i}.tar.gz chr${i}.bed chr${i}.bim chr${i}.fam chr${i}.log # store previous as compressed files
plink --bfile chr${i}.sampfilter --make-bed --out chr${i}.sampfilter.maf005 --maf 0.05 | tee >> log4.filter.maf.txt
tar --remove-files -czvf chr${i}.sampfilter.tar.gz chr${i}.sampfilter.bed chr${i}.sampfilter.bim chr${i}.sampfilter.fam chr${i}.sampfilter.log chr${i}.sampfilter.nosex # store previous as compressed files
 done

 
echo "complete" >> log3.filter.samples.txt
echo "complete" >> log4.filter.maf.txt


#########################################################
###############Harmonization
#########################################################
mkdir -p $HARMONIZED
cd $HARMONIZED

### Harmonize KORA based on Popgen
# Uses linkage disequilibrium patterns to determine the correct strand of SNPS

for i in `seq 1 22` # chromossome x is left behind because there is no availability in KORA dataset
 do

java -jar /work_ifs/sukmb447/apps/GenotypeHarmonizer-1.4.23/GenotypeHarmonizer.jar --input $KORA/chr${i}.merged.sampfilter.maf005 -id --ref $POPGEN/chr${i}.sampfilter.maf005 --update-reference-allele --output kora_chr${i}

 

### Make major allele the reference and filter SNPs missing in < 10% of Popgen
plink2 --bfile $POPGEN/chr${i}.sampfilter.maf005 --geno 0.1 --make-bed --maj-ref --max-alleles 2 --out popgen_chr${i}.filtered

 

### write reference alleles to file
cat popgen_chr${i}.filtered.bim | awk '{print $2, $5}' > popgen_chr${i}.ref

 

### Filter KORA: remove SNPs < 10% missingness AND make PopGen reference Allele also ref in KORA
plink2 --bfile kora_chr${i} --geno 0.1 --make-bed --ref-allele popgen_chr${i}.ref --max-alleles 2 --out kora_chr${i}.filtered

 

### Write list of all SNPs available in BOTH cohorts
cat *chr${i}.filtered.bim | awk '{print $2}' | sort | uniq -c | awk '{if($1==2) print $2}' > chr${i}.keep

### Merge cohorts; careful: this is plink version 1 now!
plink --bfile popgen_chr${i}.filtered --bmerge kora_chr${i}.filtered --out PopGen_and_KORA.harmonized.chr${i} --make-bed --extract chr${i}.keep

 done 


echo "complete"

# Compact non-merged files

for i in `seq 1 22` # chromossome x is left behind because there is no availability in KORA dataset
 do
 tar --remove-files -czvf kora_chr${i}.tar.gz kora_chr${i}.b* kora_chr${i}.fam kora_chr${i}.log
 tar --remove-files -czvf kora_chr${i}.filtered.tar.gz kora_chr${i}.filtered.* # store previous as compressed files
 tar --remove-files -czvf popgen_chr${i}.filtered.tar.gz popgen_chr${i}.filtered*  # store previous as compressed files
 done 


echo "complete"


#### MAKE PCA

mkdir -p $PCA
cd $PCA

# The PCA needs to be based on the whole genotype dataset. Therefore, I will first merge the dataset

for i in {1..22}
do
echo ${HARMONIZED}/PopGen_and_KORA.harmonized.chr${i} >> mergelist.txt
done

plink --merge-list mergelist.txt --make-bed --out PopGen_and_KORA.harmonized.all

# Now, get the pca
plink --bfile PopGen_and_KORA.harmonized.all --pca --out PopGen_and_KORA.harmonized.all.pca

### GET MAF for whole dataset

mkdir -p $MAF
cd $MAF

# Copy combined cohort from PCA folder
ln -s $PCA/PopGen_and_KORA.harmonized.all* $MAF/

# Calculate MAF for whole cohort

plink --bfile PopGen_and_KORA.harmonized.all --freqx --out PopGen_and_KORA.harmonized.all.MAF
